### ISSUE TITLE
Error when downloading multiple stickers file: 429 : Too Many Requests

### ISSUE THREAD

**Issue by vellrya** [tdlib/td#1517]
Hello.

I use tdlib in my android app to allow users add sticker pack from their telegram account.

Some code fragments:

```
//already resolved by setName before
val tgStickerSet = client.send(TdApi.GetStickerSet(tgSetId))
for (stickerObj in tgStickerSet.stickers) {
        println("process sticker: $i")
	val stickerFilePath = if (stickerObj.sticker.local.isDownloadingCompleted && stickerObj.sticker.local.path.isNotEmpty()) stickerObj.sticker.local.path
		else {
			val stickerFileDownload = (client.send(TdApi.DownloadFile(stickerObj.sticker.id,10, 0, 0, true)) as TdApi.Object)
			if (stickerFileDownload is TdApi.Error) {
				//error here :(
			} else {
				(stickerFileDownload as TdApi.File).local.path
			}
		}
	//add info to local db omitted
	if (stickerSet.animated) {
		val thumbFilePath = if (stickerObj.thumbnail!!.file.local.isDownloadingCompleted && stickerObj.thumbnail!!.file.local.path.isNotEmpty()) stickerObj.thumbnail!!.file.local.path
			else {
				val thumbFileDownload = (client.send(TdApi.DownloadFile(stickerObj.thumbnail!!.file.id,10, 0, 0, true)) as TdApi.Object)
				if (thumbFileDownload is TdApi.Error) {
					//log error
				} else {
					(thumbFileDownload as TdApi.File).local.path
				}
			}
	}
}
if (tgStickerSet.thumbnail!=null) {
	val thumbFilePath = if (tgStickerSet.thumbnail!!.file.local.isDownloadingCompleted && tgStickerSet.thumbnail!!.file.local.path.isNotEmpty()) tgStickerSet.thumbnail!!.file.local.path
	else {
		val setThumbFileDownload = (client.send(TdApi.DownloadFile(tgStickerSet.thumbnail!!.file.id,10, 0, 0, true)) as TdApi.Object)
		if (setThumbFileDownload is TdApi.Error) {
			//log error
		} else {
			(setThumbFileDownload as TdApi.File).local.path
		}
	}
}
```

Shortly, this code gets info about sticker set, then loading all sticker file to user device, preview of animated sticker and set thumbnail (if available).
This works perfectly, if user add sets one by one manually, but for new user it would be nice to give him the ability to initially sync all his sets with his Telegram account. And this is where the problems begin. 
Most users have 10-20 sets added, and if I try to load them all automatically (not in parallel, just one after another without pauses), pretty quickly I start getting the following errors:

> [25.04.2021 13:12:33.243] process sticker: 0
> [ 2][t 3][1619345553.245043516][AuthDataShared.cpp:106][#1][!Downloader]	DcId{4} [auth_key_id:3447461370262914863][state:OK][created_at:1618776901.000000]
> [ 2][t 3][1619345553.245634317][Session.cpp:169][#1][!SessionProxy:4:download_small#0]	Generate new session_id 3403942066884336125 for auth key 3447461370262914863 for DC-4
> [25.04.2021 13:12:33.483] process sticker: 1
> [ 2][t 3][1619345553.485265016][Session.cpp:169][#1][!SessionProxy:4:download#0]	Generate new session_id 2531027305756814225 for auth key 3447461370262914863 for DC-4
> [25.04.2021 13:12:33.666] process sticker: 2
> [25.04.2021 13:12:33.754] process sticker: 3
> ...
> [25.04.2021 13:12:37.177] process sticker: 43
> [25.04.2021 13:12:37.177] process sticker: 44
> [ 2][t 3][1619345562.240763902][Session.cpp:794][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955036208126967328][error_code:-503][msg:Timeout]
> [ 2][t 5][1619345562.242085695][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:20316161][tl:0xb15a9afc][state:Query]] [timeout:1.000000][total_timeout:1.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.120:443] to MEDIA DcId{4} from [0.0.0.0:48004]
> [ 2][t 3][1619345568.298850059][Session.cpp:794][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955036234174920468][error_code:-503][msg:Timeout]
> [ 2][t 5][1619345568.299229860][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:20316161][tl:0xb15a9afc][state:Query]] [timeout:2.000000][total_timeout:3.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.120:443] to MEDIA DcId{4} from [0.0.0.0:48004]
> [ 2][t 3][1619345575.355047464][Session.cpp:794][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955036264480760220][error_code:-503][msg:Timeout]
> [ 2][t 5][1619345575.355656862][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:20316161][tl:0xb15a9afc][state:Query]] [timeout:4.000000][total_timeout:7.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.120:443] to MEDIA DcId{4} from [0.0.0.0:48004]
> [ 2][t 3][1619345584.415250539][Session.cpp:794][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955036303397162992][error_code:-503][msg:Timeout]
> [ 2][t 5][1619345584.415567398][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:20316161][tl:0xb15a9afc][state:Query]] [timeout:8.000000][total_timeout:15.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.120:443] to MEDIA DcId{4} from [0.0.0.0:48004]
> [ 2][t 3][1619345597.474468231][Session.cpp:794][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955036359481552284][error_code:-503][msg:Timeout]
> [ 2][t 5][1619345597.474937915][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:20316161][tl:0xb15a9afc][state:Query]] [timeout:16.000000][total_timeout:31.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.120:443] to MEDIA DcId{4} from [0.0.0.0:48004]
> [ 2][t 3][1619345618.542109727][Session.cpp:794][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955036449931496192][error_code:-503][msg:Timeout]
> [ 2][t 5][1619345618.542459964][NetQueryDelayer.cpp:72][#1][!NetQueryDelayer]	Failed: [Query:[id:20316161][tl:0xb15a9afc][state:Query]] [timeout:32.000000][total_timeout:63.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.120:443] to MEDIA DcId{4} from [0.0.0.0:48004]
> [ 2][t 5][1619345618.544512510][FileManager.cpp:3663][#1][!FileManager]	Failed to Download file 1422(0) of type stickers: [Error : 429 : Too Many Requests: retry after 32]
> Error { code = 400 message = "File download has failed or was cancelled" } 

This is log from server app, but it doesn't matter. I didn't use the app for a few days, then three different sets in a row were successfully loaded (not in parallel, one by one, if that matters), and on the 4th set I got this error.
Further attempts to load this set after some time were also unsuccessful, the error always appeared on "process sticker: 44".

It turns out that there are some hard limits for downloading files? I don't re-download files if they were downloaded earlier, I use local result and try not to load servers unnecessarily.

Is this normal behavior and I'll just have to put up with such strict limits, or maybe I'm doing something wrong and there is still a possibility to download several sets of stickers in a row?

Thanks for reading such a long text, I hope for your advice :)

**Comment by levlam** [comment]
The shown error means that there is something wrong with a specific file, not with your code or something else. It should be caused by a specific sticker. Do you have remote.id of the sticker, which fails to download with this error?

**Comment by vellrya** [comment]
```
2021-04-26 15:59:33.388 ? I/System.out: Sticker {
2021-04-26 15:59:33.388 ? I/System.out:   setId = 439863097025363974
2021-04-26 15:59:33.388 ? I/System.out:   width = 512
2021-04-26 15:59:33.388 ? I/System.out:   height = 371
2021-04-26 15:59:33.388 ? I/System.out:   emoji = "🍕"
2021-04-26 15:59:33.388 ? I/System.out:   isAnimated = false
2021-04-26 15:59:33.388 ? I/System.out:   isMask = false
2021-04-26 15:59:33.388 ? I/System.out:   maskPosition = null
2021-04-26 15:59:33.388 ? I/System.out:   thumbnail = Thumbnail {
2021-04-26 15:59:33.388 ? I/System.out:     format = ThumbnailFormatWebp {
2021-04-26 15:59:33.388 ? I/System.out:     }
2021-04-26 15:59:33.388 ? I/System.out:     width = 128
2021-04-26 15:59:33.388 ? I/System.out:     height = 93
2021-04-26 15:59:33.388 ? I/System.out:     file = File {
2021-04-26 15:59:33.388 ? I/System.out:       id = 3014
2021-04-26 15:59:33.388 ? I/System.out:       size = 3672
2021-04-26 15:59:33.388 ? I/System.out:       expectedSize = 3672
2021-04-26 15:59:33.388 ? I/System.out:       local = LocalFile {
2021-04-26 15:59:33.388 ? I/System.out:         path = ""
2021-04-26 15:59:33.388 ? I/System.out:         canBeDownloaded = true
2021-04-26 15:59:33.388 ? I/System.out:         canBeDeleted = false
2021-04-26 15:59:33.388 ? I/System.out:         isDownloadingActive = false
2021-04-26 15:59:33.388 ? I/System.out:         isDownloadingCompleted = false
2021-04-26 15:59:33.388 ? I/System.out:         downloadOffset = 0
2021-04-26 15:59:33.388 ? I/System.out:         downloadedPrefixSize = 0
2021-04-26 15:59:33.388 ? I/System.out:         downloadedSize = 0
2021-04-26 15:59:33.388 ? I/System.out:       }
2021-04-26 15:59:33.389 ? I/System.out:       remote = RemoteFile {
2021-04-26 15:59:33.389 ? I/System.out:         id = "AAMCBAADFQABYIU-pF6Tx-OopJIy1rHXhKCrMm0AAmQCAAIotRoGpXF5GAiGsHl9yIwwAAQBAAdtAAPGIwACHgQ"
2021-04-26 15:59:33.389 ? I/System.out:         uniqueId = "AQADfciMMAAExiMAAg"
2021-04-26 15:59:33.389 ? I/System.out:         isUploadingActive = false
2021-04-26 15:59:33.389 ? I/System.out:         isUploadingCompleted = true
2021-04-26 15:59:33.389 ? I/System.out:         uploadedSize = 3672
2021-04-26 15:59:33.389 ? I/System.out:       }
2021-04-26 15:59:33.389 ? I/System.out:     }
2021-04-26 15:59:33.389 ? I/System.out:   }
2021-04-26 15:59:33.389 ? I/System.out:   sticker = File {
2021-04-26 15:59:33.389 ? I/System.out:     id = 3015
2021-04-26 15:59:33.389 ? I/System.out:     size = 18078
2021-04-26 15:59:33.389 ? I/System.out:     expectedSize = 18078
2021-04-26 15:59:33.389 ? I/System.out:     local = LocalFile {
2021-04-26 15:59:33.389 ? I/System.out:       path = ""
2021-04-26 15:59:33.389 ? I/System.out:       canBeDownloaded = true
2021-04-26 15:59:33.389 ? I/System.out:       canBeDeleted = false
2021-04-26 15:59:33.389 ? I/System.out:       isDownloadingActive = false
2021-04-26 15:59:33.389 ? I/System.out:       isDownloadingCompleted = false
2021-04-26 15:59:33.389 ? I/System.out:       downloadOffset = 0
2021-04-26 15:59:33.389 ? I/System.out:       downloadedPrefixSize = 0
2021-04-26 15:59:33.389 ? I/System.out:       downloadedSize = 0
2021-04-26 15:59:33.389 ? I/System.out:     }
2021-04-26 15:59:33.389 ? I/System.out:     remote = RemoteFile {
2021-04-26 15:59:33.389 ? I/System.out:       id = "CAACAgQAAxUAAWCFPqRek8fjqKSSMtax14SgqzJtAAJkAgACKLUaBqVxeRgIhrB5HgQ"
2021-04-26 15:59:33.389 ? I/System.out:       uniqueId = "AgADZAIAAii1GgY"
2021-04-26 15:59:33.389 ? I/System.out:       isUploadingActive = false
2021-04-26 15:59:33.389 ? I/System.out:       isUploadingCompleted = true
2021-04-26 15:59:33.389 ? I/System.out:       uploadedSize = 18078
2021-04-26 15:59:33.389 ? I/System.out:     }
2021-04-26 15:59:33.389 ? I/System.out:   }
2021-04-26 15:59:33.389 ? I/System.out: }
2021-04-26 15:59:38.449 ? W/DLTD: [ 2][t 3][1619441978.449477910][Session.cpp:791][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955450316444000196][error_code:-503][msg:Timeout]
2021-04-26 15:59:38.449 ? W/DLTD: [ 2][t 4][1619441978.449842929][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:19857409][tl:0xb15a9afc][state:Query]] [timeout:1.000000][total_timeout:1.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.111:443] to MEDIA DcId{4} from [192.168.88.11:37240]
2021-04-26 15:59:44.512 ? W/DLTD: [ 2][t 3][1619441984.511957168][Session.cpp:791][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955450342479867708][error_code:-503][msg:Timeout]
2021-04-26 15:59:44.512 ? W/DLTD: [ 2][t 4][1619441984.512795925][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:19857409][tl:0xb15a9afc][state:Query]] [timeout:2.000000][total_timeout:3.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.111:443] to MEDIA DcId{4} from [192.168.88.11:37240]
2021-04-26 15:59:51.576 ? W/DLTD: [ 2][t 3][1619441991.576249122][Session.cpp:791][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955450372823047372][error_code:-503][msg:Timeout]
2021-04-26 15:59:51.576 ? W/DLTD: [ 2][t 4][1619441991.576618194][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:19857409][tl:0xb15a9afc][state:Query]] [timeout:4.000000][total_timeout:7.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.111:443] to MEDIA DcId{4} from [192.168.88.11:37240]
2021-04-26 16:00:00.649 ? W/DLTD: [ 2][t 3][1619442000.649098157][Session.cpp:791][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955450411749858620][error_code:-503][msg:Timeout]
2021-04-26 16:00:00.649 ? W/DLTD: [ 2][t 4][1619442000.649316310][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:19857409][tl:0xb15a9afc][state:Query]] [timeout:8.000000][total_timeout:15.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.111:443] to MEDIA DcId{4} from [192.168.88.11:37240]
2021-04-26 16:00:13.733 ? W/DLTD: [ 2][t 3][1619442013.733805179][Session.cpp:791][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955450467901761064][error_code:-503][msg:Timeout]
2021-04-26 16:00:13.734 ? W/DLTD: [ 2][t 4][1619442013.734092235][NetQueryDelayer.cpp:82][#1][!NetQueryDelayer]	Delay: [Query:[id:19857409][tl:0xb15a9afc][state:Query]] [timeout:16.000000][total_timeout:31.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.111:443] to MEDIA DcId{4} from [192.168.88.11:37240]
2021-04-26 16:00:34.800 ? W/DLTD: [ 2][t 3][1619442034.800825119][Session.cpp:791][#1][!Session:4:download_small#0]	Session::on_message_result_error from mtproto [id:6955450558466234124][error_code:-503][msg:Timeout]
2021-04-26 16:00:34.801 ? W/DLTD: [ 2][t 4][1619442034.801254987][NetQueryDelayer.cpp:72][#1][!NetQueryDelayer]	Failed: [Query:[id:19857409][tl:0xb15a9afc][state:Query]] [timeout:32.000000][total_timeout:63.000000] because of [Error : -503 : Timeout] from Session:4:download_small#0::Connect::Tcp::[149.154.165.111:443] to MEDIA DcId{4} from [192.168.88.11:37240]
2021-04-26 16:00:34.801 ? W/DLTD: [ 2][t 4][1619442034.801611185][FileManager.cpp:3585][#1][!FileManager]	Failed to Download file 3015(0) of type stickers: [Error : 429 : Too Many Requests: retry after 32]
2021-04-26 16:00:34.803 ? I/System.out: Error {
2021-04-26 16:00:34.803 ? I/System.out:   code = 400
2021-04-26 16:00:34.803 ? I/System.out:   message = "File download has failed or was cancelled"
2021-04-26 16:00:34.803 ? I/System.out: }
```

This is sticker from setName=pussheen (resolved id: 439863097025363974)
I checked set in official client and found that some stickers are also not loaded and in their place is empty.
If this is file problem, why I get "Error : 429 : **Too Many Requests**" in debug logs, is it safe to ignore this type of error, or what should I do in this situation?)

**Comment by levlam** [comment]
If you receive 429, you can try to redownload the file in a few hours. If it is still unavailable after a few such tries, then you can give up and mark the file as unavailable.
